source/include/microapp/cs_MicroappProtocol.h

```diff
@@ -19,158 +19,237 @@ typedef struct {
-		 */
-		int _cocounter;
-
-		/**
-		 * Implementing count down for a coroutine counter.
-		 */
-		int _coskip;
-
-		/**
-		 * Store data for callbacks towards the microapp.
-		 */
-		CircularBuffer<uint8_t>* _callbackData;
-
-		/**
-		 * Max number of mesh messages that will be queued.
-		 */
-		const uint8_t MAX_MESH_MESSAGES_BUFFERED = 3;
-
-		/**
-		 * Buffer received mesh messages.
-		 *
-		 * Starts with microapp_buffered_mesh_message_header_t, followed by the message.
-		 */
-		CircularBuffer<microapp_buffered_mesh_message_t> _meshMessageBuffer;
-
-	protected:
-
-		/**
-		 * Call the loop function (internally).
-		 */
-		void callLoop(int & cntr, int & skip);
-
-		/**
-		 * Initialize memory for the microapp.
-		 */
-		uint16_t initMemory();
-
-		/**
-		 * Load ram information, set by microapp.
-		 */
-		uint16_t interpretRamdata();
-
-		/**
-		 * Handle a received mesh message.
-		 */
-		void onMeshMessage(MeshMsgEvent event);
-
-	public:
-		static MicroappProtocol& getInstance() {
-			static MicroappProtocol instance;
-			return instance;
-		}
-
-		/**
-		 * Set IPC ram data.
-		 */
-		void setIpcRam();
-
-		/**
-		 * Actually run the app.
-		 */
-		void callApp(uint8_t appIndex);
-
-		/**
-		 * Call setup and loop functions.
-		 */
-		void callSetupAndLoop(uint8_t appIndex);
-
-		/**
-		 * Receive events (for example for i2c)
-		 */
-		void handleEvent(event_t & event);
-
-		/**
-		 * Handle mesh command.
-		 */
-		void handleMeshCommand(microapp_mesh_header_t* meshCommand, uint8_t* payload, size_t payloadSize);
+class MicroappProtocol : public EventListener {
+private:
+	/**
+	 * Singleton, constructor, also copy constructor, is private.
+	 */
+	MicroappProtocol();
+	MicroappProtocol(MicroappProtocol const&);
+	void operator=(MicroappProtocol const&);
+
+	/**
+	 * Local flag to check if app did boot. This flag is set to true after the main of the microapp is called
+	 * and will be set after that function returns. If there is something wrong with the microapp it can be used
+	 * to disable the microapp.
+	 *
+	 * TODO: Of course, if there is something wrong, _booted will not be reached. The foolproof implementation
+	 * sets first the state to a temporarily "POTENTIAL BOOT FAILURE" and if it after a reboots finds the state
+	 * at this particular step, it will disable the app rather than try again. The app has to be then enabled
+	 * explicitly again.
+	 */
+	bool _booted;
+
+	/**
+	 * Local flag to indicate that ram section has been loaded.
+	 */
+	bool _loaded;
+
+	/**
+	 * Debug mode
+	 */
+	bool _debug;
+
+	/**
+	 * Address to setup() function.
+	 */
+	uintptr_t _setup;
+
+	/**
+	 * Address to loop() function.
+	 */
+	uintptr_t _loop;
+
+	/**
+	 * Addressees of pin interrupt service routines.
+	 */
+	pin_isr_t _pin_isr[MAX_PIN_ISR_COUNT];
+
+	/**
+	 * Addressees of ble interrupt service routines.
+	 */
+	ble_isr_t _ble_isr[MAX_BLE_ISR_COUNT];
```

Comment:
Though it doesn't look better in this case, stick to the code style =]
```suggestion
	ble_isr_t _bleIsr[MAX_BLE_ISR_COUNT];
```